### ISSUE TITLE
ON/OFF Toggle Widget

### ISSUE THREAD

**Issue by jerryn70** [julian-klode/dns66#67]
Adding ON/OFF Toggle Widget will help users to quickly switch ON and OFF adblocking.  Then you might think why switching OFF adblocking , the answer is some applications won't work when vpn is enabled example 'shareit' . So we are forced to turn OFF vpn to use that apps. At this point toggle widget if helpful

**Comment by debagos** [comment]
>some applications won't work when vpn is enabled
You can blacklist apps in DNS66.

**Comment by jerryn70** [comment]
@debagos That won't work. We need to to completely stop vpn to make that app work

**Comment by jarettmillard** [comment]
A quick settings tile would be useful as well.

**Comment by smichel17** [comment]
If you are only interested in being able to pause the app while it is running, I'd suggest adding an action to the notification.

If you want to be able to cold start as well, you could (in addition to a widget, since this is api 24+ only) support a quick settings tile.

That's done by adding this to your manifest:

    <service
        android:name=".receiver.TileReciever"
        android:icon="@drawable/toggle_icon"
        android:label="@string/app_name"
        android:permission="android.permission.BIND_QUICK_SETTINGS_TILE">
        <intent-filter>
            <action android:name="android.service.quicksettings.action.QS_TILE" />
        </intent-filter>
    </service>

and then you add a class (in this case, TileReciever) that inherits from [TileService()](https://developer.android.com/reference/android/service/quicksettings/TileService.html), and override the `onClick()` method.

**Comment by julian-klode** [comment]
@smichel17 PR #62 has an implementation of a quick settings tile (which is issue #26).

We can't have an action in the notification. Well we could, but it would automatically reconnect on connectivity changes (there's only one paused state - the waiting for network state). Quick settings and widgets have the advantage of not needing a second pause state.

That said, this all depends on the work in the hostdb branch - currently, pausing anything is a fairly wasteful affair (we re-read all hosts files after unpausing...). And well, Android is buggy, so even though I don't need to re-read host files, I'll probably have to add a 2 second sleep before establishing a new VPN connection, as Android otherwise goes mad and no packets reach the service anymore...

With the new refactoring in hostdb brach, we could easily have more states, as the service can be controlled more accurately now - by sending it commands via a file descriptor. So we could easily have a PAUSE_FULLY command.

**Comment by smichel17** [comment]
The way we do it in [Red Moon](https://github.com/raatmarien/red-moon) (a screen filter) is:

Action in the notification to stop the filter.
When the filter is stopped, stop the service running in the foreground, and send a non-sticky notification.
When the filter is stopped in the background, do the same, but dismiss the notification.

So, we don't add any additional state for that, it's either on or off, but a user can stop the service and then start it again (provided they don't swipe the notification away while it's stopped).

**Comment by Redoregon** [comment]
Brief user input: the irritation of notification screen clutter greatly outweighs the ability to quickly pause something that remains running the vast majority of the time.


-------- Original Message --------

**Comment by smichel17** [comment]
@redoregon DNS66's notification can be disabled in its menu. Although I agree, that notification clutter is annoying.

**Comment by Redoregon** [comment]
Yes, of course as long as the notification remains deselectable, all is well :-)


-------- Original Message --------

**Comment by julian-klode** [comment]
I would not call a lowest-priority notification clutter. It's always at the end of the notification list, and not even shown on the lock screen. If you have about 3 other notifications, you'd have to scroll down to see it. The real reason to have it is to prevent Android from killing the service (well, to make it less likely).

**Comment by smichel17** [comment]
Also, the android system can disable any notification. Maybe instead of allowing the user to turn off the notification via the app, we should have a link to the place in the android system where they can turn off notifications from DNS66. That way we can always startForeground(), and users can still hide the notification.

**Comment by julian-klode** [comment]
@smichel17 The notification thing is implemented in 86af494
